packages/compass-data-modeling/src/store/diagram.ts

```diff
@@ -223,9 +253,128 @@ export const diagramReducer: Reducer<DiagramState> = (
       updatedAt: new Date().toISOString(),
     };
   }
+  if (isAction(action, DiagramActionTypes.COLLECTION_SELECTED)) {
+    return {
+      ...state,
+      selectedItems: { type: 'collection', id: action.namespace },
+    };
+  }
+  if (isAction(action, DiagramActionTypes.RELATIONSHIP_SELECTED)) {
+    return {
+      ...state,
+      selectedItems: {
+        type: 'relationship',
+        id: action.relationship.id,
+      },
+    };
+  }
+  if (isAction(action, SidePanelActionTypes.EDIT_RELATIONSHIP_FORM_CANCELED)) {
+    if (state.selectedItems?.type === 'relationship') {
+      return {
+        ...state,
+        // Just unselect when canceling edit on relationship, this is basically
+        // the same as closing the drawer in this case
+        selectedItems: null,
+      };
+    }
+    return state;
+  }
+  if (isAction(action, SidePanelActionTypes.SIDE_PANEL_CLOSE_CLICKED)) {
+    return {
+      ...state,
+      selectedItems: null,
+    };
+  }
+  if (isAction(action, SidePanelActionTypes.EDIT_RELATIONSHIP_FORM_SUBMITTED)) {
+    if (
+      state.selectedItems?.type === 'relationship' &&
+      state.selectedItems.id === action.relationshipId
+    ) {
+      return {
+        ...state,
+        selectedItems: null,
+      };
+    }
+    return state;
+  }
   return state;
 };
 
+export function selectCollection(
+  namespace: string
+): DataModelingThunkAction<Promise<void>, CollectionSelectedAction> {
+  return async (dispatch, getState) => {
+    const { diagram, sidePanel } = getState();
+
+    if (
+      diagram?.selectedItems?.type === 'collection' &&
+      diagram.selectedItems.id === namespace
+    ) {
+      return;
+    }
+
+    if (
+      sidePanel.viewType === 'relationship-editing' &&
+      sidePanel.relationshipFormState.modified
+    ) {
+      const allowSelect = await showConfirmation({
+        title: 'You have unsaved chages',
+        description:
+          'You are currently editing a relationship. Are you sure you want to select a different collection?',
+      });
+      if (!allowSelect) {
+        return;
+      }
+    }
```

Comment:
I'm not sure this behavior is needed, but when playing around with the feature it seemed like a good idea to add it, otherwise it's easy to loose your form state by clicking around (not sure many people will do this though)